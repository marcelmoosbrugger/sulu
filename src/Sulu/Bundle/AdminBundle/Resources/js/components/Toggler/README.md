The toggler is an boolean input element and has no internal state. It has to be managed from the outside,
like shown in the following example:

```
initialState = {checked: false};
onChange = (checked) => setState({checked});
<Toggler checked={state.checked} onChange={onChange} />
```
