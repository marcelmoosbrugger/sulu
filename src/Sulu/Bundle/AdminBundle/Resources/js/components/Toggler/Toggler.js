// @flow
import React from 'react';
import type {ElementRef} from 'react';
import togglerStyles from './toggler.scss';

type Props = {
    checked: boolean,
    onChange?: (checked: boolean) => void,
}

export default class Toggler extends React.PureComponent<Props> {
    input: ElementRef<'input'>;

    handleChange = (event: SyntheticEvent<HTMLInputElement>) => {
        if (this.props.onChange) {
            this.props.onChange(event.currentTarget.checked);
        }
    };

    handleClick = () => this.input.click();
    handleInputClick = (event: Event) => event.stopPropagation();
    setInput = (input: ElementRef<'input'>) => this.input = input;

    render() {
        return (
            <span onClick={this.handleClick} className={togglerStyles.toggler}>
                <input
                    ref={this.setInput}
                    type="checkbox"
                    checked={this.props.checked}
                    onClick={this.handleInputClick}
                    onChange={this.handleChange} />
                <span />
            </span>
        );
    }
}
