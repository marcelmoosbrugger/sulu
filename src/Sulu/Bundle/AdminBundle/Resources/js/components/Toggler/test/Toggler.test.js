/* eslint-disable flowtype/require-valid-file-annotation */
import {shallow, render} from 'enzyme';
import React from 'react';
import Toggler from '../Toggler';

test('The component should render in unchecked state', () => {
    const toggler = render(<Toggler checked={false} />);
    expect(toggler).toMatchSnapshot();
});

test('The component should render in checked state', () => {
    const toggler = render(<Toggler checked={true} />);
    expect(toggler).toMatchSnapshot();
});

test('A click on the toggler should trigger the change callback', () => {
    const onChangeSpy = jest.fn();
    const toggler = shallow(<Toggler checked={false} onChange={onChangeSpy} />);
    toggler.find('input').simulate('change', {currentTarget: {checked: true}});
    expect(onChangeSpy).toHaveBeenCalledWith(true);
    toggler.find('input').simulate('change', {currentTarget: {checked: false}});
    expect(onChangeSpy).toHaveBeenCalledWith(false);
});
