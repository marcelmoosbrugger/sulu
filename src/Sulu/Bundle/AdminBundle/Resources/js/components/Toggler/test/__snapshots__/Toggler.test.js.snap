// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The component should render in checked state 1`] = `
<span
  class="toggler"
>
  <input
    checked=""
    type="checkbox"
  />
  <span />
</span>
`;

exports[`The component should render in unchecked state 1`] = `
<span
  class="toggler"
>
  <input
    type="checkbox"
  />
  <span />
</span>
`;
